packages/client/src/client/Web/serviceWorkerRegistration.ts

```diff
@@ -53,51 +56,78 @@ export function register(config?: Config) {
         })
       } else {
         // Is not localhost. Just register service worker
-        registerValidSW(swUrl, config)
+        registerServiceWorker(swUrl, config)
       }
     })
   }
 }
 
-function registerValidSW(swUrl: string, config?: Config) {
+function addStateChangeHandler(
+  serviceWorkerInstance: ServiceWorker,
+  registration: ServiceWorkerRegistration,
+  registrationConfig: Config | undefined,
+) {
+  serviceWorkerInstance.onstatechange = () => {
+    console.debug(
+      "'installing' SW changed state to: ",
+      serviceWorkerInstance.state,
+    )
+    if (serviceWorkerInstance.state === "installed") {
+      if (navigator.serviceWorker.controller) {
+        // At this point, the updated precached content has been fetched,
+        // but the previous service worker will still serve the older
+        // content until all client tabs are closed.
+        console.debug(
+          "New content is available and will be used when all " +
+            "tabs for this page are closed. See https://cra.link/PWA.",
+        )
+        serviceWorkerInstance.postMessage({ type: "SKIP_WAITING" })
+      } else {
+        // At this point, everything has been precached.
+        // It's the perfect time to display a
+        // "Content is cached for offline use." message.
+        console.debug("Content is cached for offline use.")
+
+        // Execute callback
+        if (registrationConfig && registrationConfig.onSuccess) {
+          registrationConfig.onSuccess(registration)
+        }
+      }
+    } else if (serviceWorkerInstance.state === "activated") {
+      if (registrationConfig && registrationConfig.onUpdate) {
+        registrationConfig.onUpdate(registration)
```

Comment:
the real "done" state is activated .. at this point the new SW is controlling the current page - or it is capable of taking control of the latest modules in the event of a refresh .. any other state, and the previous activated SW will be in charge .. 
There has been a lot more logging around here to verify all this .. what is left now is the bare minimum for us to see if there are any other edge/corner cases around this problematic lifecycle !